src/kernel-install/ukify.py

```diff
@@ -0,0 +1,187 @@
+
+    # Suppress printing of usage synopsis on errors
+    p.error = lambda message: p.exit(2, f'{p.prog}: error: {message}\n')
+
+    p.add_argument('linux',
+                   type=pathlib.Path)
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
+
+    p.add_argument('--os-release',
+                   type=pathlib.Path)
+
+    p.add_argument('--devicetree',
+                   type=pathlib.Path)
+    p.add_argument('--splash',
+                   type=pathlib.Path)
+
+    p.add_argument('--efi-arch',
+                   choices=('ia32', 'x64', 'arm', 'aa64', 'riscv64'))
+
+    p.add_argument('--stub',
+                   type=pathlib.Path)
+
+    p.add_argument('--output', '-o',
+                   nargs='?')
+
+    p.add_argument('--key',
+                   type=pathlib.Path)
+    p.add_argument('--cert',
+                   type=pathlib.Path)
+
+    p.add_argument('--measure',
+                   action='store_true')
+
+    args = p.parse_args()
+
+    if bool(args.key) ^ bool(args.cert):
+        raise ValueError('--key and --cert must be specified together')
+
+    if args.os_release is None:
+        p = pathlib.Path('/etc/os-release')
+        if not p.exists():
+            p = pathlib.Path('/usr/lib/os-release')
+        args.os_release = p
+
+    if args.efi_arch is None:
+        args.efi_arch = guess_efi_arch()
+
+    if args.stub is None:
+        args.stub = f'/usr/lib/systemd/boot/efi/linux{args.efi_arch}.efi.stub'
+
+    if args.output is None:
+        suffix = '.efi' if args.key else '.unsigned.efi'
+        args.output = args.linux.name + suffix
+
+    return args
+
+
+def check_inputs(args):
+    for name, value in vars(args).items():
+        if name == 'output':
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        open(value).close()
+
+
+def make_uki(args):
+    files = []  # a holder for temporary files
+    sections = []
+
+    def add_section(name, content, offset):
+        if not content:
+            return
+
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            files.append(tmp)
+            content = tmp.name
+
+        sections.append((name, content, offset))
+
+    add_section('osrel',   args.os_release,    0x20_000)
+    add_section('cmdline', args.cmdline,       0x30_000)
+    add_section('dtb',     args.devicetree,    0x40_000)
+    add_section('splash',  args.splash,       0x100_000)
+    add_section('linux',   args.linux,      0x2_000_000)
+    add_section('initrd',  args.initrd,     0x3_000_000)
+
+    if args.key:
+        unsigned = tempfile.NamedTemporaryFile(prefix='linux')
+        output = unsigned.name
+    else:
+        output = args.output
+
+    objcopy = [
+        'objcopy',
```

Comment:
Have you checked that this works for arched other than x86? Historically, not all arches have PE support in binutils (hence why those ship the elf stub too). And some tools have issues with the gnu-efi created PE binaries: if I remember correctly, sbsign doesn't like our arm binaries, and llvm-binutils definitely doesn't.

Reply:
No, I haven't really tested this at all. Nevertheless, this essentially duplicates what dracut does and our docs say. I'm not even sure if I could test this properly. I have a rpi with 64 bits, but it has it's special bootloader IIRC, and wouldn't boot an UKI. Though we should definitely add some CI for this. Even if it turns out that we do the tool differently, it'd be useful to have tests. They can be adapted to a different scheme later.